Add barrier covers to zwave_js integration

kpine opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Implements the cover platform for motorized barrier devices for the zwave_js integration.

This adds cover implementation for Z-Wave motorized barrier devices (such as a garage door opener).

The "barrier signal state" is also exposed as a binary switch. This allows turning on/off the visual or audible signals. The behavior will vary by model.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16554

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

kpine committed:
Add barrier covers to zwave_js integration

Adds cover support for motorized barrier devices. This only considers
the "Secure Barrier Add-on" device types.

Also adds binary switches to toggle the visual/audible signals.

MartinHjelmare commented on homeassistant/components/zwave_js/switch.py:
```diff
@@ -62,3 +71,59 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+    @property
+    def is_on(self) -> Optional[bool]:  # type: ignore
+        """Return a boolean for the state of the switch."""
+        return bool(self._state)
```
This won't return None now, so the return value isn't optional. If we want it optional we should not use the `bool` copy.

kpine replied:
If None is returned, does that result in an "unknown" state for the entity? Maybe that is preferable.

MartinHjelmare replied:
Not currently, but it may change in the future. I think it's more correct to return None since we don't know the state.

kpine commented:
Some questions I had about the implementation.

#### Signaling switches

- These values are intentionally **not** refreshed by zwave-js after setting them. I chose to go the route where HA optimistically assumes it succeeded, because it probably always will unless there was an error. In that case though, does the API allow for detecting if a command failed, so I could revert the state if it did? Or, would it be preferable to schedule a refresh (poll) manually after the set and let the state be updated normally (if we have that capability yet)? The downside of the optimistic choice is that something like zwavejs2mqtt will be out of state w/o a refresh.
- Should these signaling switches be disabled by default? They have never been implemented before in the other integrations. I added them because I thought it would be useful to expose the full functionality, but perhaps they are "advanced" entities.

#### Barrier position

zwave-js exposes the barrier position as a separate value. Even if the device does not actually report exact position (0-100%), zwave-js will emulate it and report 0% for closed state, and 100% for open state. But the barrier CC does not implement a "set position" command, it's merely a report (and optional at that). I think that means I can't implement the property, even for attributes, because the cover entity [assumes](https://github.com/home-assistant/core/blob/eb0d1bb673230b8bbece6019aa639dc55477160c/homeassistant/components/cover/__init__.py#L220) existence of the property means set is supported? Originally I wanted to report this but removed it, I don't think I can include it really.

#### Barrier state

Besides closed/opening/open/closing, there is a stopped state. HA does not support a stopped state. I see some integrations convert stopped to open/closed. However, the z-wave spec is very specific on what stopped means:

![image](https://user-images.githubusercontent.com/791794/107679562-90a7d100-6c51-11eb-88c1-1d5dd66338e4.png)

The only way to get out of a stopped state is with an "open" command. Ideally, in this state and w/o HA supporting such a state, I would report it as open instead, because that's the most accurate state. However, in order to issue an open cover, I don't think HA will allow the user to do so if it's in the open state already, and I would not want to report a false state that the door is closed. Is there any way to implement this (hasn't been done in previous integrations)?

MartinHjelmare commented:
I suggest returning `None` for `is_closed` if the cover is stopped before reaching either end position. That will be translated to unknown state and the user should be able to open the cover. We can add a guard check in the close_cover method that raises an error if it's called when the state is unknown / None. Then the user will be notified of this.

MartinHjelmare commented:
Setting a value (sending a command) will raise `FailedCommand` if it failed. It may also raise `NotConnected` if it was called when the client isn't connected.

https://github.com/home-assistant-libs/zwave-js-server-python/blob/master/zwave_js_server/exceptions.py

MartinHjelmare commented:
It's ok to update state optimistically, I think. If the command fails it will raise before we update state.

kpine commented on homeassistant/components/zwave_js/cover.py:
```diff
@@ -20,7 +21,14 @@
 LOGGER = logging.getLogger(__name__)
-SUPPORT_GARAGE = SUPPORT_OPEN | SUPPORT_CLOSE
+
+BARRIER_TARGET_CLOSE = 0
+BARRIER_TARGET_OPEN = 255
+
+BARRIER_STATE_CLOSED = 0
+BARRIER_STATE_CLOSING = 252
+BARRIER_STATE_OPENING = 254
+BARRIER_STATE_OPEN = 255
```
These should make their way into the python lib.

kpine committed:
Fix type errors

kpine committed:
Return None for a stopped state

kpine commented:
> 
> 
> I suggest returning `None` for `is_closed` if the cover is stopped before reaching either end position. That will be translated to unknown state and the user should be able to open the cover. We can add a guard check in the close_cover method that raises an error if it's called when the state is unknown / None. Then the user will be notified of this.

This worked pretty well. When in the stopped state both of the close and open buttons in the UI are available. The open command succeeds.

I decided not to add the exception and just allow the command go through. The device will ignore it if it's not the right one. That way I don't have to assume anything. Let me know if you prefer otherwise.

kpine committed:
Update discovery after rebase

jcam commented:
Nortek/GoControl GD00Z-8-GC currently has no door state or control, though I can control the sound and lights!
dump attached. I suppose it's possible that it isn't working because the opener isn't installed on the door yet, though I can see the door open/close status flipping in the zwave-js logs when I turn the door sensor.

[gd00z-8.json.txt](https://github.com/home-assistant/core/files/5974535/gd00z-8.json.txt)

kpine commented:
Yeah, so in node-zwave-js 6.2.0, the specific device type for the openers is `"UNKNOWN (0x07)"`. The master branch has the desired value `"Secure Barrier Add-on"`, I _could_ update this PR to support 6.2.0 if we want to.

```python
        device_class_specific={
            "Secure Barrier Add-on",
            "UNKNOWN (0x07)",
        },
```

jcam commented:
Re-ran with node-zwave-js @master. Now it sees the barrier operator! Unfortunately the state reads unknown, despite showing stopped & 0% in the zwavejs2mqtt screen
[gd00z-8.json.txt](https://github.com/home-assistant/core/files/5974797/gd00z-8.json.txt)

jcam commented:
Ah, this is because it doesn't know where the door is, because it isn't on the door.
The position sensor does seem to be missing though?

kpine committed:
Fix required_values schema search

The search for required values was exiting before searching all of the
node's values, therefore if the first value did not match the required
scheming, it returned not found.

kpine committed:
Relax detection of motorized barrier devices

Previous OZW-based integrations were either very strict (only Secure
Barrier Add-on) or were looser (Entry Control + tons of seemingly random
command classes).

In reality, all the really matters is the existence of the Barrier
Command class and the ability to get and set the barrier status.

In the future there might be Open/Close-only devices. Those can be
specialized later.

kpine commented:
> 
> 
> Ah, this is because it doesn't know where the door is, because it isn't on the door.
> 
> I triggered the up operation and used the door tilt and it worked exactly as it should :)

Correct, since HA doesn't provide a "stopped" state, this is the best solution as suggested by @MartinHjelmare . If you take a look at the state diagram I posted earlier, you'll see the Barrier Operator CC spec dictates that a barrier in the stopped state can only go to the open state, thus we need to allow for the open_cover command. If the cover was in the "open" state instead, open_cover would not be available. If it was in the "closed" state, it would basically be a lie.

MartinHjelmare commented on homeassistant/components/zwave_js/cover.py:
```diff
@@ -99,3 +111,57 @@ async def async_stop_cover(self, **kwargs: Any) -> None:
+        # issuing an open cover command. Return None in this case which
+        # produces an unknown state and allows it to be resolved with an open
+        # command.
+        if self.info.primary_value.value == BARRIER_STATE_STOPPED:
```
It would be good to add a test for this case as we expect this to happen whenever the cover is stopped.

kpine committed:
Add test for stopped state

kpine committed:
Check device class

kpine commented on homeassistant/components/zwave_js/discovery.py:
```diff
@@ -78,7 +78,7 @@ class ZWaveDiscoverySchema:
     device_class_specific: Optional[Set[str]] = None
     # [optional] additional values that ALL need to be present on the node for this scheme to pass
-    required_values: Optional[Set[ZWaveValueDiscoverySchema]] = None
+    required_values: Optional[List[ZWaveValueDiscoverySchema]] = None
```
The type is not hashable.

kpine commented on homeassistant/components/zwave_js/discovery.py:
```diff
@@ -362,6 +374,24 @@ class ZWaveDiscoverySchema:
     ),
+    # cover
+    # motorized barriers
+    ZWaveDiscoverySchema(
+        platform="cover",
+        hint="motorized_barrier",
+        primary_value=ZWaveValueDiscoverySchema(
+            command_class={CommandClass.BARRIER_OPERATOR},
+            property={"currentState"},
+            type={"number"},
+        ),
+        required_values=[
+            ZWaveValueDiscoverySchema(
+                command_class={CommandClass.BARRIER_OPERATOR},
+                property={"targetState"},
+                type={"number"},
+            ),
+        ],
+    ),
```
Removed the check for device classes. This allows the cover to support node-zwave-js v6.2.0 and any other potential motorized barrier types that we don't know about. The functionality isn't dependent on the types, only the barrier class.

There are some specific types that could result in different HA device classes (e.g. DEVICE_CLASS_GATE), but I haven't seen this kind of device yet, so just assume garage.

kpine commented on homeassistant/components/zwave_js/discovery.py:
```diff
@@ -430,13 +460,10 @@ def async_discover_values(node: ZwaveNode) -> Generator[ZwaveDiscoveryInfo, None
-                for val_scheme in schema.required_values:
-                    for val in node.values.values():
-                        if not check_value(val, val_scheme):
-                            required_values_present = False
-                            break
```
This was returning False for the very first value that didn't match, even if the value matched later on.

kpine commented:
Just another observation... I implemented `current_cover_position`, but noticed when the door is opening or closing the cover entity controls are disabled in the lovelace UI. This is because my GDC, and probably the other ones on the market, don't return position information, and so zwave-js only reports 0 or 100% (fully closed or open). The [frontend code](https://github.com/home-assistant/frontend/blob/d4d38a880df741071e7485ef4fabc9e14cdee42b/src/components/ha-cover-controls.ts#L80-L100)  disables the close/open buttons if the cover position of fully open (100) or closed (0). If not implemented, it checks the is_opening/is_closing properties. In a stopped state ("unknown") it allows both controls to be enabled.

There might be ways to work around it (modify position while it changes state?), but it doesn't seem worth the hassle.

kpine commented:
Marked this as ready for review since it functions with current versions of node-zwave-js.

We can wait on a version bump for zwave-js-server-python if you'd like me to incorporate the barrier constants that were added, otherwise I can follow up with another change.

kpine commented:
TODO: update docs to remove references about garage doors not being supported. <https://www.home-assistant.io/integrations/zwave_js/#current-limitations>

MartinHjelmare reviewed (approved):
Looks good!
